### PR TITLE
Refactor Volatility Models

### PR DESCRIPTION
**Pull request by Martin-Molinero**
#### Description

- Removing usages of `Security.Configuration` properties from existing
Volatility Models, without modifying existing interfaces. Adding unit
tests for existing behavior
- Adding new `BaseVolatilityModel`, which will have a setter for a
`ISubscriptionDataConfigProvider` that will be called by the system.
*Maintaining retro compatibility*. Adding unit tests for this behavior.
- Adding reduced interface `ISubscriptionDataConfigProvider`
- Adding new `SubscriptionDataConfigExtensions` static class which will
provide methods used to determine different configuration properties for
a given set of `SubscriptionDataConfigs`. The behavior was extracted
from current `Security` class behavior regarding the `SubscriptionBag`.
Adding unit tests covering expected behavior here.
#### Related Issue

Closes https://github.com/QuantConnect/Lean/issues/2633
#### Motivation and Context

https://github.com/QuantConnect/Lean/projects/5
#### Requires Documentation Change

N/A
#### How Has This Been Tested?

Unit tests
#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Common/Data/SubscriptionDataConfigExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        /// <summary>
+        /// Extension method used to determine if FillForward is enabled
+        /// for a given set of <see cref="SubscriptionDataConfig"/>
+        /// </summary>
+        /// <param name="subscriptionDataConfigs"></param>
+        /// <returns>True, at least one subscription has it enabled</returns>
+        public static bool FillForward(
+            this IEnumerable<SubscriptionDataConfig> subscriptionDataConfigs)
+        {
+            return subscriptionDataConfigs.Any(x => x.FillDataForward);
```
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Common/Data/SubscriptionDataConfigExtensions.cs]: Yes, good point, note I'm just maintaining existing behavior here. I'll rename 👍

File: Common/Python/VolatilityModelPythonWrapper.cs [hunk 2]
```diff
@@ -68,12 +70,28 @@ public void Update(Security security, BaseData data)
...
+        /// <summary>
+        /// Sets the <see cref="ISubscriptionDataConfigProvider"/> instance to use.
+        /// </summary>
+        /// <param name="subscriptionDataConfigProvider">Provides access to registered <see cref="SubscriptionDataConfig"/></param>
+        public override void SetSubscriptionDataConfigProvider(
+            ISubscriptionDataConfigProvider subscriptionDataConfigProvider)
+        {
+            using (Py.GIL())
+            {
+                if (_model.HasAttr("SetSubscriptionDataConfigProvider"))
```
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Common/Python/VolatilityModelPythonWrapper.cs]: Correct, I'm adding this to avoid breaking existing user models

File: Common/Securities/Volatility/BaseVolatilityModel.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+using System.Collections.Generic;
+using QuantConnect.Data;
+using QuantConnect.Interfaces;
+
+namespace QuantConnect.Securities.Volatility
+{
+    /// <summary>
+    /// Represents a base model that computes the volatility of a security
+    /// </summary>
+    public abstract class BaseVolatilityModel : IVolatilityModel
```
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Common/Securities/Volatility/BaseVolatilityModel.cs]: Good catch! Will do 👍

File: Common/Securities/Volatility/IVolatilityModel.cs [hunk 4]
```diff
@@ -17,12 +17,14 @@
...
 using System.Collections.Generic;
 using System.Linq;
+using QuantConnect.Securities.Volatility;
 
 namespace QuantConnect.Securities
 {
     /// <summary>
     /// Represents a model that computes the volatility of a security
     /// </summary>
+    /// <remarks><see cref="BaseVolatilityModel"/></remarks>
```
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Common/Securities/Volatility/IVolatilityModel.cs]: Will do 👍

File: Common/Securities/Volatility/RelativeStandardDeviationVolatilityModel.cs [hunk 5]
```diff
@@ -109,25 +113,37 @@ public void Update(Security security, BaseData data)
...
-                new HistoryRequest(utcStartTime, 
+                throw new Exception(
+                    "RelativeStandardDeviationVolatilityModel.GetHistoryRequirements(): " +
+                    "SubscriptionDataConfigProvider was not set."
+                );
+            }
+            var configurations = SubscriptionDataConfigProvider
+                .GetSubscriptionDataConfigs(security.Symbol)
+                .ToList();
+            var configuration = configurations.First();
```
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Common/Securities/Volatility/RelativeStandardDeviationVolatilityModel.cs]: True, in this case, the first SDC is used to obtain `.Symbol` and `.DataTimeZone`. Primarily the objective was to avoid doing `MarketHoursDatabase.FromDataFolder().GetDataTimeZone()` which did not seem to be required given the data was in context.

File: Engine/AlgorithmManager.cs [hunk 6]
```diff
@@ -984,16 +985,27 @@ private IEnumerable<TimeSlice> Stream(IAlgorithm algorithm, IDataManager dataMan
...
+            Log.Trace("ProcessVolatilityHistoryRequirements(): Updating volatility models with historical data...");
 
             foreach (var kvp in algorithm.Securities)
             {
                 var security = kvp.Value;
 
                 if (security.VolatilityModel != VolatilityModel.Null)
                 {
+                    // start: this is a work around to maintain retro compatibility
+                    var baseType = security.VolatilityModel as BaseVolatilityModel;
```
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Engine/AlgorithmManager.cs]: Good idea, will do 👍

File: Common/Data/SubscriptionDataConfigExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        /// <param name="subscriptionDataConfigs"></param>
+        /// <returns>The first DataNormalizationMode,
+        /// <see cref="DataNormalizationMode.Adjusted"/> if there  are no subscriptions</returns>
+        public static DataNormalizationMode DataNormalizationMode(
+            this IEnumerable<SubscriptionDataConfig> subscriptionDataConfigs)
+        {
+            return subscriptionDataConfigs.
+                Select(x => x.DataNormalizationMode)
+                .DefaultIfEmpty(QuantConnect.DataNormalizationMode.Adjusted)
+                .First(); ;
```
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Common/Data/SubscriptionDataConfigExtensions.cs]: Yes, this is definitely something to keep in mind. I think in each case, the call site should determine which should be the desired behavior, something to dig into 👍

[the remark this answers is not in the document]
Comment (jaredbroad) [reply Common/Data/SubscriptionDataConfigExtensions.cs]: The `security.SetDataNormalizationMode()` is commonly used; long term will need a standard for determining which subscription is updated vs deprecating it. Say explicitly the ManualUniverse Subscription?

[the remark this answers is not in the document]
Comment (mchandschuh) [reply Common/Data/SubscriptionDataConfigExtensions.cs]: RE: `security.SetDataNormalizationMode` -- this is fine to remain in my opinion as part of the API/algorithm layer -- the concern is more focused around the engine making decisions pertaining to the normalization mode in the scope of a security vs the scope of a single subscription. _Even_ if the API enforces one normalization mode per security (which it doesn't, only the `Security` object does (loosely)), the underlying engine code is written in such a way that it _appears_ to not care (nor should it, it's simply a projection applied to an isolated stream of data) if a security's subscriptions are normalized under different modes

*TL;DR* -- only point is to be careful defining values at the 'security' level and if possible, try to force the call sites requiring these values to operate the logic per subscriptions vs per security -- and _if_ it truly is a per security context, perhaps consider why normalization mode is needed for that logic.

File: Common/Securities/Volatility/RelativeStandardDeviationVolatilityModel.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mchandschuh) [reply Common/Securities/Volatility/RelativeStandardDeviationVolatilityModel.cs]: hmm, interesting dilemma indeed. The system as written actually _assumes_ all of a security's data have the same market hours entries (by virtue of the market-hours-db and it's consumers) -- so if that's the only value being pulled from the config than I'd say that's completely safe 👍
